docs/src/modules/components/Demo.js

```diff
@@ -39,56 +59,93 @@ function getDemoName(location) {
-    product = 'base';
-    name = 'MUI Base';
-  }
-  if (asPathWithoutLang.startsWith('/x/')) {
-    name = 'MUI X';
-  }
+  const { canonicalAs } = pathnameToLanguage(router.asPath);
+
+  return React.useMemo(() => {
+    let product;
+    let name = 'Material UI';
+    if (canonicalAs.startsWith('/joy-ui/')) {
+      product = 'joy-ui';
+      name = 'Joy UI';
+    } else if (canonicalAs.startsWith('/base/')) {
+      product = 'base';
+      name = 'MUI Base';
+    } else if (canonicalAs.startsWith('/x/')) {
+      name = 'MUI X';
+    }
 
-  const title = `${getDemoName(githubLocation)} demo — ${name}`;
-  if (codeVariant === CODE_VARIANTS.TS && demo.rawTS) {
     return {
-      codeVariant: CODE_VARIANTS.TS,
-      githubLocation: githubLocation.replace(/\.js$/, '.tsx'),
-      language: userLanguage,
-      raw: demo.rawTS,
-      Component: demo.tsx,
-      sourceLanguage: 'tsx',
-      title,
+      scope: demo.scope,
+      jsxPreview: demo.jsxPreview,
+      ...(codeVariant === CODE_VARIANTS.TS && demo.rawTS
+        ? {
+            codeVariant: CODE_VARIANTS.TS,
+            githubLocation: githubLocation.replace(/\.js$/, '.tsx'),
+            raw: demo.rawTS,
+            Component: demo.tsx,
+            sourceLanguage: 'tsx',
+          }
+        : {
+            codeVariant: CODE_VARIANTS.JS,
+            githubLocation,
+            raw: demo.raw,
+            Component: demo.js,
+            sourceLanguage: 'jsx',
+          }),
+      title: `${getDemoName(githubLocation)} demo — ${name}`,
       product,
+      language: userLanguage,
     };
-  }
-
-  return {
-    codeVariant: CODE_VARIANTS.JS,
-    githubLocation,
-    language: userLanguage,
-    raw: demo.raw,
-    Component: demo.js,
-    sourceLanguage: 'jsx',
-    title,
-    product,
-  };
+  }, [canonicalAs, codeVariant, demo, githubLocation, userLanguage]);
 }
 
-// TODO: replace with React.useOpaqueReference if it is released
-function useUniqueId(prefix) {
-  // useOpaqueReference
-  const [id, setId] = React.useState();
-  React.useEffect(() => {
-    setId(Math.random().toString(36).slice(2));
-  }, []);
+function useDemoElement({ demoOptions, demoData, usePreview, code, debouncedSetError }) {
+  // Memoize to avoid rendering the demo more than it needs to be.
+  // For example, avoid a render when the demo is hovered.
+  return React.useMemo(() => {
+    // No need for a live environment if the toolbar is hidden.
+    if (demoOptions.hideToolbar) {
+      return <demoData.Component />;
+    }
+
+    const codeToRun = usePreview
+      ? trimLeadingSpaces(demoData.raw).replace(trimLeadingSpaces(demoData.jsxPreview), code)
+      : code;
+
+    // In production mode, it's better to hydrate the demo as soon as possible (`fallback`).
+    // Later on, we can accept live edits.
+    if (process.env.NODE_ENV === 'production') {
+      return (
+        <ForwardProps>
+          {(forwardProps) => (
+            <NoSsr fallback={<demoData.Component />}>
```

Comment:
I think we can use `IntersectionObserver` to make it better

Reply:
This can make sense for a follow-up. I suspect we could make it even better by only rendering `ReactRunner` when the code to run is different from the initial code. It